ch3/exercise3-10/comma.go

```diff
@@ -11,12 +12,27 @@ func main() {
 	}
 }
 
-//!+
-// comma inserts commas in a non-negative decimal integer string.
+
 func comma(s string) string {
+	var buf bytes.Buffer
 	n := len(s)
 	if n <= 3 {
 		return s
 	}
-	return comma(s[:n-3]) + "," + s[n-3:]
+
+	offset := n % 3
+	if offset > 0 {
+		buf.WriteString(s[0:offset])
+		buf.WriteByte(',')
+	}
+	s = s[offset:]
+	
+	for i := len(s); i > 0; i -= 3{
+		buf.WriteString(s[i-3 : i])
+		if i > 3 {
+			buf.WriteByte(',')
+		}
+		
+	}
```

Comment:
This logic is incorrect.  `i := len(s)` means you're starting at the end of the string, and going backwards (after the initial offset).  Look at your testing output:

`12345678` becomes   `12,678,345`

Reply:
oops, silly mistake. fixed!